rfd/00xx-pod-access-requests.md

```diff
@@ -0,0 +1,555 @@
+Currently, Teleport appends the RBAC principals to Teleport Roles or to the user
+itself. It means that if a user wants temporary access to a single Kubernetes POD,
+the cluster administrator has to create a Kubernetes RBAC Role, assign it to a
+Kubernetes RBAC principal and create a role with the principal. After that, the user
+has to request access to the temporary role. Once someone approves the role access
+request, the user can impersonate it.
+
+In addition to the complex process, limiting access to pods using Kubernetes RBAC
+has limitations because Kubernetes RBAC does not support controlling access resources
+using pattern matching.
+The RBAC Role must include the resource name to grant access, but it requires
+knowing the Pod name in advance. In many cases, it is impossible to know them
+because Pods generated from Kubernetes objects like Deployments include random
+strings in their names and change each time the Pod is deleted or relocated from
+one node to the other.
+
+This RFD proposes an extension to Teleport RBAC where it will support
+requesting access to Pods - including Pods names following a pattern - without
+relying on requesting access to Teleport Roles.
+
+## Role definition
+
+This RFD proposes an alternative to limit or extend user access to certain pods
+in a Kubernetes cluster from Teleport.
+
+To achieve it, it proposes changes to Teleport's Role spec to introduce a new
+field `kubernetes_pods` where it's possible to specify the Pod names to allow
+access.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: my-kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_pods: ["A","B","C"]
+```
+
+This is the first iteration and the goal is to share how it would work. We can
+extend the API to limit certain actions while accessing pods, e.g., a user can list
+and see the logs of a Pod but cannot exec into it or copy any of its contents.
+This might be useful under debug situations where Pod A connects to Pod B and
+the user can ask for total access to Pod A, but only list and log access to Pod B.
+
+## Details
+
+We investigated two different approaches to extend the Teleport's RBAC to control
+access to specific Pods. The ultimate goal of both approaches is to allow users
+to temporarily request access to certain Pods.
+
+The two approaches are considerably different, and we will detail each one to
+briefly describe the implementation details and their limitations.
+
+1. Teleport RBAC controls access to each Pod.
+
+When a user has at least one role that has a non-empty list of `kubernetes_pods`,
+Teleport will only allow access to the Pods included in the list, even for cases
+where other roles allow unrestricted access to other Pods or the Kubernetes RBAC
+allows access to other Pods. `kubernetes_pods` is a subset of the super-set allowed
+by the user's Kubernetes RBAC principals.
```

Comment:
Maybe worth clarifying that this is the case for roles that match the cluster your connecting to. So for example of you have:

```
roleA:
  kube_labels: env=prod
  kube_pods: x, y, z

roleB:
  kube_labels: env=stage
```

Then you'll still be able to access anything in "env=stage" clusters and only pods x, y, z in "env=prod" clusters.

Meaning, kube_pods is not a "poison pill" and doesn't apply globally to any roles you have.